clorm/orm/core.py

```diff
@@ -2360,115 +2357,223 @@ def __iter__(self):
 # tuple. Otherwise simply returns the value.
 # ------------------------------------------------------------------------------
 
-def _preprocess_field_value(field_defn, v):
-    predicate_cls = field_defn.complex
-    if not predicate_cls: return v
-    mt = predicate_cls.meta
-    if isinstance(v, predicate_cls): return v
-    if (mt.is_tuple and isinstance(v,Predicate) and v.meta.is_tuple) or \
-       isinstance(v, tuple):
-        if len(v) != len(mt):
-            raise ValueError(("mis-matched arity between field {} (arity {}) and "
-                             " value (arity {})").format(field_defn, len(mt), len(v)))
+def _instance_from_tuple(predicate_cls, v):
+    if isinstance(v, tuple) or (isinstance(v,Predicate) and v.meta.is_tuple):
         return predicate_cls(*v)
-    else:
-        return v
+    raise TypeError(f"Cannot map {v} ({type(v)}) to {predicate_cls.meta} object")
 
 # ------------------------------------------------------------------------------
 # Helper functions for PredicateMeta class to create a Predicate
 # class constructor.
 # ------------------------------------------------------------------------------
 
-# Construct a Predicate via an explicit (raw) clingo.Symbol object
-def _predicate_init_by_raw(self, **kwargs):
-    raw = kwargs["raw"]
-    self._raw = raw
+AnySymbol = Union[clingo.Symbol,noclingo.Symbol]
+
+def _get_symbols():
+    return symbols
```

Comment:
Yes, exactly. 

But it is ugly and I'm going to change this. The way I've integrated the noclingo stuff is not clean so I'm going to try to make it a bit more self-contained within the noclingo module. The idea is that `noclingo.Function()`, `noclingo.String()`, etc will internally decide what type of symbol to create based on the mode so all the indirection is contained within these functions and we don't need hackiness like `_get_symbols()`. Which will also make it easier if/when noclingo can be removed.

Reply:
I've removed most of the hackiness of things like the `_get_symbols()` function. However, I think there is only one linger issue.

Now, the overhead introduced by the flexibility of noclingo is confined to `noclingo.py` and it is simply to add an indirection when calling the various symbol creating functions. This overhead is pretty small and you can only see the performance difference when you look at averages over multiple runs of the test code. But it is still annoying because there are many/most occasions when you don't need this functionality so why have the overhead?.

I've added some if statements in the module that could enable/disable noclingo when the module is loaded. I think the default should be to have it disabled.  But I'm not sure what's the most pythonic way to trigger this. You would need to set some flag before the module is loaded to tell it to enable noclingo. Or potentially do it using a environment variable, in a similar way to `PYTHONHASHSEED`.  

Any thoughts?

Reply:
What about using something like a context-manager if you want to use the noclingo-mode?
I'm not sure if it is possible but you could switch to the noclingo-mode when entering the ctxt-manager and switch back to clingo-mode when leaving the ctxt-manager?

Reply:
That's a nice idea, and might be a good way to switch to noclingo-mode. But it's not what I had in mind here, where I want to completely disable it, so there is no way to switch modes. We could get a small performance improvement because we can remove the indirection of the various symbol creation functions. 

Anyway, for the moment I won't disable it and will just merge the branch. Can revisit this later.